# MovieLens

##DUE APRIL 21, 2016

[MovieLens](http://www.movielens.org/) is a website where users can submit ratings for movies that they watch and receive recommendations for other movies they might enjoy. The data is collected and made publicly available for research. We will be working with a data set of 1 million user ratings of movies. You can find this data set and even larger ones at http://grouplens.org/datasets/movielens/.


## Reading in the Data

Note that the data consists of three data frames: one with information about the users, another containing the ratings, and yet another with information about the movies. See the readme file (/data/movielens/README) for more information.

In [1]:
import pandas as pd

unames = ['user_id', 'gender', 'age', 'occupation', 'zip']
users = pd.read_table('/data/movielens/users.dat', sep='::', header=None,
                      names=unames, engine="python")

rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_table('/data/movielens/ratings.dat', sep='::', header=None,
                        names=rnames, engine="python")

mnames = ['movie_id', 'title', 'genres']
movies = pd.read_table('/data/movielens/movies.dat', sep='::', header=None,
                       names=mnames, engine="python")

## Question 1

Develop a way to rank movies. Why is it not a good idea to simply sort the movies by average rating? (You may want to try calculating this first.) Then, explain your methodology and use it to produce a list of the Top 10 movies of all time.

In [2]:
#ratings

In [3]:
# YOUR CODE HERE
#raise NotImplementedError()

# ratings_movie_id_avg = ratings.groupby(["movie_id" ]).mean()#.get(["rating"])
# #This would calculate the mean of each column, including userid and timestamp for ea movieid
# ratings_movie_id_avg




#Another way to rank movies:
# A 'top' movie is a movie with not only THOUSANDS of ratings, but on average, people gave it a 5 star rating.

num_ratings_per_movie = ratings.groupby(["movie_id"])[['rating']].count()
num_ratings_per_movie = num_ratings_per_movie.ix[num_ratings_per_movie['rating'] >= 2000 ] # get movies with >= # of ratings
num_ratings_per_movie = num_ratings_per_movie.reset_index() # call this after groupby
num_ratings_per_movie = num_ratings_per_movie.merge( ratings, on= 'movie_id')

top_movies = pd.merge(num_ratings_per_movie,users, on = "user_id")
top_movies = top_movies.ix[ top_movies['rating_y'] == 5].groupby('movie_id')[['rating_y']].mean()
top_movies.columns = ['average_rating']
top_movies.head(n=10)




# avg_ratings = ratings.ix[ratings['rating'] >= 4, :].groupby('movie_id')[ ['rating'] ].mean()
# # don't use get('rating') as that will use all cols when calling .mean() on it
# avg_ratings

average_rating
movie_id                
1                      5
110                    5
260                    5
296                    5
318                    5
356                    5
480                    5
527                    5
589                    5
593                    5

It doesn't make sense to sort the movies by its average ratings alone because there may be some movies where only a few people gave it a rating. A movie with a rating defined by only a few reviewers is not a good rating because a few reviewers is not reprsentative of the entire audience at all.

The table outputted shows the top 10 movies of all time.

## Question 2

Restrict to movies with at least 200 ratings. For each movie, calculate the difference between the average male rating and the average female rating. Based on these differences between average male ratings and average female ratings, what movies were the most male-friendly? What movies were the most female-friendly?

In [4]:
# YOUR CODE HERE
#raise NotImplementedError()


num_ratings_per_movie = ratings.groupby(["movie_id"])[['rating']].count()
num_ratings_per_movie = num_ratings_per_movie.ix[num_ratings_per_movie['rating'] >= 200]
num_ratings_per_movie = num_ratings_per_movie.reset_index()
num_ratings_per_movie = num_ratings_per_movie.merge( ratings, on= 'movie_id')
#num_ratings_per_movie
#next step: merge num_ratings_per_movie_200 with original ratings table and then filter out.


data_merged_rate = pd.merge(num_ratings_per_movie,users, on = "user_id")

# # inner join:
data_merged_rate['Avg Female Rating'] = data_merged_rate.ix[ data_merged_rate['gender'] == 'F' ].groupby(['movie_id']).mean()['rating_y']
data_merged_rate['Avg Female Rating'] = data_merged_rate['Avg Female Rating'].fillna(0)
data_merged_rate['Avg Male Rating'] = data_merged_rate.ix[ data_merged_rate['gender'] == 'M' ].groupby( ['movie_id'] ).mean()['rating_y']
data_merged_rate['Avg Male Rating'] = data_merged_rate['Avg Male Rating'].fillna(0)
# #['rating'] is necessary because I only want to add a column

data_merged_rate['Diff Between Females\' and Males\' Avg Ratings' ] = data_merged_rate['Avg Female Rating']\
     - data_merged_rate['Avg Male Rating']
data_merged_rate = data_merged_rate.sort_values(by = 'Diff Between Females\' and Males\' Avg Ratings' ,ascending = True)

#data_merged_rate

print(data_merged_rate.head(n = 3))
#for most male friendly
print("=========================================================================")
print(data_merged_rate.tail(n = 3))
# for most  female friendly

      movie_id  rating_x  user_id  rating_y  timestamp gender  age  \
1201      3744       421       19         4  978147373      M    1   
3760      1784      1424       65         5  977888204      M   35   
1194      3624       791       19         5  978146984      M    1   

      occupation    zip  Avg Female Rating  Avg Male Rating  \
1201          10  48073           3.494949         4.221300   
3760          12  55803           2.878788         3.555147   
1194          10  48073           2.944444         3.585227   

      Diff Between Females' and Males' Avg Ratings  
1201                                     -0.726351  
3760                                     -0.676359  
1194                                     -0.640783  
      movie_id  rating_x  user_id  rating_y  timestamp gender  age  \
2468      1721      1546       38         4  978044771      F   18   
203       2699      1367        8         5  978229347      M   25   
1088      1792       448       19         3 

Movies(movie_id) that were most male friendly : 3744, 1784, 3624

Movies(move_id) that were most female friendly: 1721, 2699, 1792

*note: If difference between females' and males' average ratings is positive, that means that on average, females gave a particular movie higher rating than males did (female_friendly). If difference is negative, males gave it higher ratings on average.

## Question 3

Calculate the average rating by genre. Note that a movie can belong to multiple genres. You will have to write some code that parses the `genres` column of the `movies` table. What genre had the highest average rating? What genre had the lowest?

In [19]:
# YOUR CODE HERE
#raise NotImplementedError()
#avg_ratings = ratings.ix[ratings['rating'] >= 4, :].groupby('movie_id')[ ['rating'] ].mean()



split_genres = movies[ 'genres'].str.cat(sep = '|') # returns a string
#split_genres

split_genres = split_genres.split("|") # must concat before doing this step
split_genres

distinct_genres_list = []

#create list of distinct genres

for genre in split_genres:
 
    
    if genre not in distinct_genres_list:
        
        distinct_genres_list.append(genre)



data_merged_genre = pd.merge(ratings, movies, on = 'movie_id')
data_merged_genre = data_merged_genre.sort_index()
data_merged_genre

for genre in distinct_genres_list:
    
    if genre in data_merged_genre['genres']: #never true
        data_merged_genre[genre] = True
        
data_merged_genre
        
#.str.contains('Drama')
#data_merged_genre['genres'].str.contains('Drama')
#use .split?

# data_merged_genre = data_merged_genre.groupby('genres' )[ ['rating'] ].mean()
# print(data_merged_genre.sort_values(by = 'rating').tail())
# print(data_merged_genre.sort_values(by = 'rating').head(n = 50))

user_id  movie_id  rating   timestamp  \
0              1      1193       5   978300760   
1              2      1193       5   978298413   
2             12      1193       4   978220179   
3             15      1193       4   978199279   
4             17      1193       5   978158471   
5             18      1193       4   978156168   
6             19      1193       5   982730936   
7             24      1193       5   978136709   
8             28      1193       3   978125194   
9             33      1193       5   978557765   
10            39      1193       5   978043535   
11            42      1193       3   978038981   
12            44      1193       4   978018995   
13            47      1193       4   977978345   
14            48      1193       4   977975061   
15            49      1193       4   978813972   
16            53      1193       5   977946400   
17            54      1193       5   977944039   
18            58      1193       5   977933866   
19            59      1193       4   977934292   
20            62      1193       4   977968584   
21            80      1193       4   977786172   
22            81      1193       5   977785864   
23            88      1193       5   977694161   
24            89      1193       5   977683596   
25            95      1193       5   977626632   
26            96      1193       3   977621789   
27            99      1193       2   982791053   
28           102      1193       5  1040737607   
29           104      1193       2   977546620   
...          ...       ...     ...         ...   
1000179     4933      3084       3   962757020   
1000180     4802      2218       2  1014866656   
1000181     4812      2308       2   962932391   
1000182     4874       624       4   962781918   
1000183     5059      1434       4   962484364   
1000184     5947      1434       4   957190428   
1000185     5077      1868       3   962417299   
1000186     5944      1868       1   957197520   
1000187     5105       404       3   962337582   
1000188     5185       404       4   963402617   
1000189     5532       404       5   959619841   
1000190     5543       404       3   960127592   
1000191     5220      2543       3   961546137   
1000192     5754      2543       4   958272316   
1000193     5227       591       3   961475931   
1000194     5795       591       1   958145253   
1000195     5313      3656       5   960920392   
1000196     5328      2438       4   960838075   
1000197     5334      3323       3   960796159   
1000198     5334       127       1   960795494   
1000199     5334      3382       5   960796159   
1000200     5420      1843       3   960156505   
1000201     5433       286       3   960240881   
1000202     5494      3530       4   959816296   
1000203     5556      2198       3   959445515   
1000204     5949      2198       5   958846401   
1000205     5675      2703       3   976029116   
1000206     5780      2845       1   958153068   
1000207     5851      3607       5   957756608   
1000208     5938      2909       4   957273353   

                                                     title  \
0                   One Flew Over the Cuckoo's Nest (1975)   
1                   One Flew Over the Cuckoo's Nest (1975)   
2                   One Flew Over the Cuckoo's Nest (1975)   
3                   One Flew Over the Cuckoo's Nest (1975)   
4                   One Flew Over the Cuckoo's Nest (1975)   
5                   One Flew Over the Cuckoo's Nest (1975)   
6                   One Flew Over the Cuckoo's Nest (1975)   
7                   One Flew Over the Cuckoo's Nest (1975)   
8                   One Flew Over the Cuckoo's Nest (1975)   
9                   One Flew Over the Cuckoo's Nest (1975)   
10                  One Flew Over the Cuckoo's Nest (1975)   
11                  One Flew Over the Cuckoo's Nest (1975)   
12                  One Flew Over the Cuckoo's Nest (1975)   
13                  One Flew Over the Cu

Ignoring mixed genres, the Animation genre had the highest average ratings. Also, ignoring mixed genres, the Children genre had
the lowest average ratings.

## Question 4

Formulate a question of your own that you can answer using this MovieLens data. State clearly what your question is and what your findings are. Bonus points are available if you find something interesting!

**Tip:** You may find the `occupation` column of `users` to be a rich source of interesting questions. See the README file (/data/movielens/README) for information about how `occupation` is coded in the data.

In [6]:
# YOUR CODE HERE
#raise NotImplementedError()

num_ratings_by_occupation = users.merge(ratings, on = 'user_id').groupby('occupation')[['rating']].count()
num_ratings_by_occupation.columns = ['number of ratings']
num_ratings_by_occupation

number of ratings
occupation                   
0                      130499
1                       85351
2                       50068
3                       31623
4                      131032
5                       21850
6                       37205
7                      105425
8                        2706
9                       11345
10                      23290
11                      20563
12                      57214
13                      13754
14                      49109
15                      22951
16                      46021
17                      72816
18                      12086
19                      14904
20                      60397

Q: How many ratings did each group of people by occupation give? Which occupation had people that gave the most ratings? 

Users(people) with occupation 4 rated the most amount of movies as a group. One could speculate that those with occupation 4 were more willing to rate a movie that they watched. It is also possible that the movies those in occupation 4 chose to watch tended to provoke strong feelings.

## Submitting this Lab

Now, restart your kernel and re-run your entire notebook from beginning to end. Make sure there are no errors or bugs. When you have verified this, open the Terminal on JupyterHub and type 

`nbgrader submit Lab-04-14 --course dlsun`

to submit this lab.